core/src/main/scala/kafka/coordinator/transaction/TransactionLog.scala

```diff
@@ -94,18 +97,22 @@ object TransactionLog {
       new Field(TxnEntryTimestampKey, INT64, "Time the transaction was last updated"),
       new Field(TxnStartTimestampKey, INT64, "Time the transaction was started"))
 
-    private val Schemas = Map(0 -> V0)
+    private val V1 = new Schema(new Field(ProducerIdKey, INT64, "Producer id in use by the transactional id."),
+      new Field(LastProducerIdKey, INT64, "The last producer id assigned to the producer"),
+      new Field(ProducerEpochKey, INT16, "Epoch associated with the producer id"),
+      new Field(LastProducerEpochKey, INT16, "The last epoch associated with the producer id", -1.toShort),
+      new Field(TxnTimeoutKey, INT32, "Transaction timeout in milliseconds"),
+      new Field(TxnStatusKey, INT8,
+        "TransactionState the transaction is in"),
+      new Field(TxnPartitionsKey, ArrayOf.nullable(PartitionsSchema),
+        "Set of partitions involved in the transaction"),
+      new Field(TxnEntryTimestampKey, INT64, "Time the transaction was last updated"),
+      new Field(TxnStartTimestampKey, INT64, "Time the transaction was started"))
 
-    val CurrentVersion = 0.toShort
-    val Current = schemaForValue(CurrentVersion)
+    private val Schemas = Map(0 -> V0, 1 -> V1)
 
-    val ProducerIdField = V0.get(ProducerIdKey)
-    val ProducerEpochField = V0.get(ProducerEpochKey)
-    val TxnTimeoutField = V0.get(TxnTimeoutKey)
-    val TxnStatusField = V0.get(TxnStatusKey)
-    val TxnPartitionsField = V0.get(TxnPartitionsKey)
-    val TxnEntryTimestampField = V0.get(TxnEntryTimestampKey)
-    val TxnStartTimestampField = V0.get(TxnStartTimestampKey)
+    val CurrentVersion = 1.toShort
```

Comment:
I think we need to guard usage of the new schema with the inter-broker protocol version. We do something similar in `GroupMetadataManager`. 

Also, there's a tricky detail. If the IBP is on an older version, what does that mean for requests using the new version? Should they be accepted? I guess the worst case is that we wouldn't be able to detect a retry across a coordinator change. That would lead to a spurious fencing error, but I guess that's no worse than the state before this KIP. So maybe not a big deal.

Reply:
Thanks for catching that. I went with your suggestion of checking the IBP, and just accepting that we can't handle the coordinator change case on the older IBP.